mypy/checker.py

```diff
@@ -2883,7 +2883,45 @@ def conditional_type_map(expr: Expression,
         return {}, {}
 
 
-def partition_by_callable(type: Type) -> Tuple[List[Type], List[Type]]:
+def intersect_instance_callable(type: Instance, callable_type: CallableType) -> Type:
+    """Creates a fake type that represents the intersection of an
+    Instance and a CallableType.
+
+    It operates by creating a bare-minimum dummy TypeInfo that
+    subclasses type and adds a __call__ method matching callable_type."""
+
+    # Build the fake ClassDef and TypeInfo together.
+    # The ClassDef is full of lies and doesn't actually contain a body.
+    cdef = ClassDef("<callable subtype of {}>".format(type.type.name()), Block([]))
```

Comment:
This is the trickiest thing in this PR. However, there generally seems to be no better way.

There are several things that should be handled here, such as:

* The original class may be generic.
* The original class may inherit from `Tuple[...]` and in that case we should return `TupleType`, not `Instance`.
* If the original class inherits from `Any`, the new class should also do that.

Basically you should check that every `TypeInfo` attribute is initialized correctly. 

You may be able to share code with `basic_new_typeinfo` in `mypy.semanal`, but you shouldn't add a dependency from type checker to the semantic analyzer. You could move `basic_new_typeinfo` to some third module, perhaps.

Additionally, it's possible to leak these types through type inference to class and module top level scopes. This means that the types can be serialized in incremental mode, and you should test that it works. It may be necessary to store the created class in a symbol table (with a generated unique dummy name) for serialization to work, but I'm not sure about this.